db/system_keyspace.cc

```diff
@@ -1668,6 +1668,34 @@ template future<> system_keyspace::update_peer_info<sstring>(gms::inet_address e
 template future<> system_keyspace::update_peer_info<utils::UUID>(gms::inet_address ep, sstring column_name, utils::UUID);
 template future<> system_keyspace::update_peer_info<net::inet_address>(gms::inet_address ep, sstring column_name, net::inet_address);
 
+future<> system_keyspace::update_peer_info(gms::inet_address ep, data_values_map columns) {
+    if (_db.get_token_metadata().get_topology().is_me(ep)) {
+        on_internal_error(slogger, format("update_peer_info called for this node: {}", ep));
+    }
+
+    std::ostringstream ss;
+    data_values_list values;
+    values.reserve(columns.size() * 2);
+
+    if (slogger.is_enabled(log_level::debug)) {
+        slogger.debug("BEGIN BATCH");
+        for (const auto& [col, value] : columns) {
+            slogger.debug("  INSERT INTO system.{} (peer, {}) VALUES ({}, {});", PEERS, col, ep.addr(), value);
+        }
+        slogger.debug("APPLY BATCH");
+    }
+
+    ss << "BEGIN BATCH";
+    for (auto& [col, value] : columns) {
+        ss << ::format("\n  INSERT INTO system.{} (peer, {}) VALUES (?, ?);", PEERS, col);
+        values.emplace_back(ep.addr());
+        values.emplace_back(std::move(value));
+    }
+    ss << "\nAPPLY BATCH";
+
+    co_await execute_cql_query(ss.str(), values).discard_result();
+}
+
```

Comment:
I still don't like it. We have a fixed schema for the table, and a system for using prepared statements, and it gets thrown away.

We can use either a batch or prepared statements, or the UNSET thing, but not this.

We can have something like

```c++
struct batch_element {
    statements::prepared_statement::checked_weak_ptr stmt; // from prepare_internal   
    data_values_vector arguments;
};

future<> query_processor::execute_internal_batch(const std::vector<batch_element>);
```

Reply:
So with UNSET we can prepare a INSERT/UPDATE query specifying all columns and provide UNSET values for the values that are unavailable?

Reply:
Unless I misunderstood something using an unset value seems very complicated and requires highly intrusive changes.

Reply:
Why is unset difficult? You can make data_values_list accept two constructors, one from convertible_to\<data_value>, and one from variant\<data_value, unset>. The first will be used by practically everyone. The second will be used by those crazies that want to use unset.

It's true the callers will have to collect their inputs in some struct and provide it later, but it's localized to those callers. For the peers table:

```c++
struct peer_optional_values {
    std::optional<sstring> data_center;
    ...
};
```

and then the UPDATE statement will pick up values from peer_optional_values.

Reply:
Maybe I don't understand how it's supposed to work but after exploring this it looks like unset is not being handled on the modification path, only `null` is handled, and the latter deletes the cells. I guess it is only handled for the WHERE clause expressions.

Reply:
can't we just have a struct with all the fields we want to update in `update_peer_info`
```
struct peer_info_update_request {
  std::optional<inet_address> rpc_address;
  std::optional<sstring> rack;
  // ...
}
```
`update_peer_info` does `select * from peers`, merges the data with the provided `peer_info_update_request` and inserts them back with a single big static insert:
  * it statically describes the interface of `update_peer_info`, we actually don't need dynamism in this dimension;
  * we don't need all the complexities with `data_values_list`, as long as its overheads (allocation for the vector/`data_value` copies) for the data plane;
  * we can skip the insert altogether if all the values from `peer_info_update_request` are equal to the current ones;
  * simple insert allows for prepared statements